components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java

```diff
@@ -1074,7 +1075,25 @@ private void setAuthCookie(HttpServletRequest request, HttpServletResponse respo
         }
         String path = null;
         if (IdentityTenantUtil.isTenantedSessionsEnabled()) {
-            if (FrameworkUtils.isOrganizationQualifiedRequest()) {
+            String appResidentOrgId = PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                    .getApplicationResidentOrganizationId();
+            if (StringUtils.isNotBlank(appResidentOrgId)) {
+                try {
+                    String primaryOrgId = FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                            .getPrimaryOrganizationId(appResidentOrgId);
+                    tenantDomain = FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                            .resolveTenantDomain(primaryOrgId);
```

Comment:
Primary tenant domain should be available in PrivilegeCarbonContext. So we can avoid additional DB calls

Reply:
Changed